In [3]:
#!pip install --upgrade transformers datasets
#!pip install --upgrade huggingface hub
#!pip install --upgrade pip
#!pip install --upgrade torch torchvision 

In [10]:
from transformers import AutoImageProcessor, Dinov2Model
import torch, torchvision
from datasets import load_dataset

In [6]:
print(torch.__version__)
print(torchvision.__version__)

2.1.2
0.16.2


In [12]:
device = torch.device("cuda" if torch.cuda.is_available()  
                            else "mps"  if torch.backends.mps.is_available() 
                            else "cpu")
print(f"Using device: {device}")

# If you have a model, move it to the device
# model.to(device)

Using device: mps


In [13]:
#!huggingface-cli login --token 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

Token will not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.
Token is valid (permission: read).
Your token has been saved to /Users/ari/.cache/huggingface/token
Login successful


In [14]:
dataset = load_dataset("huggingface/cats-image")
image = dataset["test"]["image"][0]

image_processor = AutoImageProcessor.from_pretrained("facebook/dinov2-base")
model = Dinov2Model.from_pretrained("facebook/dinov2-base")

inputs = image_processor(image, return_tensors="pt")
with torch.no_grad():
    outputs = model(**inputs)

last_hidden_states = outputs.last_hidden_state
list(last_hidden_states.shape)

/Users/ari/Documents/Data_Science/3_semester/learning_from_las_vegas/LFLV/LasVegas/lib/python3.11/site-packages/datasets/load.py:1429: FutureWarning: The repository for huggingface/cats-image contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/huggingface/cats-image
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(
Generating test split: 1 examples [00:00,  2.24 examples/s]
config.json: 100%|██████████| 548/548 [00:00<00:00, 1.66MB/s]
model.safetensors: 100%|██████████| 346M/346M [00:28<00:00, 11.9MB/s] 


[1, 257, 768]

#### last hidden state shape

* 1 -> input batch size
* 257 -> 256 image patches the input images are split into (one additional token for a special purpose, like classification or a start/end token). 
* 768 -> This is the dimensionality of the embeddings. Each of the 257 elements (patches or tokens) is transformed into a 768-dimensional vector. 

In [15]:
last_hidden_states

tensor([[[-2.1747, -0.4729,  1.0936,  ...,  0.2041,  1.1101,  0.1363],
         [-3.2780, -0.8269, -0.9210,  ...,  1.4415, -0.5364, -0.8757],
         [-2.9129,  1.1284, -0.7306,  ...,  0.6959, -1.8791, -2.3638],
         ...,
         [-0.5463,  1.4382, -0.2563,  ...,  0.1873, -2.9950,  0.4067],
         [-3.0848,  2.0568,  1.5137,  ...,  0.9157, -2.7059,  2.2017],
         [-0.7499,  0.0903,  1.3731,  ..., -0.2961, -2.3682, -0.1329]]])